nzgmdb/data_retrieval/waveform_extraction.py

```diff
@@ -0,0 +1,1158 @@
+    trace_indexs = []
+    traces = []
+    loc = st_copy[0].stats.location
+    chan = st_copy[0].stats.channel[:2]
+
+    for unique_channel_ending in unique_channel_endings:
+        # We need to group together each component with the same ending channel str e.g. HN1 the 1
+        group = st_copy.select(
+            location=loc,
+            channel=f"{chan}{unique_channel_ending}",
+        )
+        # For each trace compute the AI and keep the highest
+        for i, tr in enumerate(group):
+            # Trim the trace to ptime_est to ptime_est + ds595 + 1std
+            ds_end_time = np.exp(ds_mean) * np.exp(ds_std)
+            tr_copy = tr.copy()
+            tr_copy.trim(ptime_est, ptime_est + ds_end_time, pad=True, fill_value=0)
+            # Get the IA and IA_norm
+            Ia = get_arias_intensity_norm(tr_copy)
+            if i == 0:
+                Ia_max = Ia[-1]
+                Ia_max_index = 0
+            else:
+                if Ia[-1] > Ia_max:
+                    Ia_max = Ia[-1]
+                    Ia_max_index = i
+        trace_indexs.append(Ia_max_index)
+        # Grab the original trace from the st and group to get the correct trace
+        group_original = st.select(
+            location=loc,
+            channel=f"{chan}{unique_channel_ending}",
+        )
+        traces.append(group_original[Ia_max_index])
+    # Check all the groups agree
+    if len(set(trace_indexs)) == 1:
+        # We combine the traces into a new stream
+        st = Stream(traces=traces)
+        trace_index = trace_indexs[0]
+        issue = False
+    else:
+        # Raise an issue here as the groups do not agree
+        issue = True
+        trace_index = None
+
+    return st, trace_index, issue
+
+
+def select_horizontal_pair(values: list[str]):
+    """
+    Selects the best horizontal channel pair from a list of available channels based on a predefined priority.
+
+    Parameters
+    ----------
+    values : list of str
+        A list of channel identifiers (e.g., ["1", "2", "X", "Y", "N", "E"]).
+
+    Returns
+    -------
+    tuple or None
+        A tuple containing the selected horizontal channel pair (e.g., ("1", "2")) or None if no valid pair is found.
+    list of str
+        A list of issues encountered during the selection process, such as incomplete pairs.
+    """
+    # Define valid pairs in priority order
+    pair_priority = [
+        ("1", "2"),  # highest priority
+        ("X", "Y"),
+        ("N", "E"),  # lowest priority
+    ]
+    found_pairs = []
+    issues = []
+    # Check each valid pair
+    for a, b in pair_priority:
+        if a in values and b in values:
+            found_pairs.append((a, b))
+    # Check for incomplete pairs (dangling values)
+    for a, b in pair_priority:
+        if (a in values) ^ (b in values):  # XOR = only one present
+            issues.append(f"Incomplete pair: missing {b if a in values else a}")
```

Comment:
If none are present, is this also an issue? If this is true, or if it is impossible for none to be present, then I might suggest a simplification of this code:

```suggestion
    for a, b in pair_priority:
        if a in values and b in values:
            found_pairs.append((a, b))
        else:
            issues.append(f"Incomplete pair: missing {b if a in values else a}")
```

Reply:
This function is only called from here where we check for more than 2 horizontal channels

`        if len(h_channels) > 2:
            # Here we need to select the correct horizontal channels`
            
I'll look at changing this

Reply:
This code does not quite do exactly what it was meant to do. Have adjusted it to make it more efficient and clear instead of multiple loops.